src/index/blockfilterindex.cpp

```diff
@@ -387,13 +389,30 @@ bool BlockFilterIndex::LookupFilter(const CBlockIndex* block_index, BlockFilter&
     return ReadFilterFromDisk(entry.pos, filter_out);
 }
 
-bool BlockFilterIndex::LookupFilterHeader(const CBlockIndex* block_index, uint256& header_out) const
+bool BlockFilterIndex::LookupFilterHeader(const CBlockIndex* block_index, uint256& header_out)
 {
+    {
+        // Check the headers cache
+        LOCK(m_cs_headers_cache);
+        auto header = m_headers_cache.find(block_index->GetBlockHash());
+        if (header != m_headers_cache.end()) {
+            header_out = header->second;
+            return true;
+        }
+    }
+
     DBVal entry;
     if (!LookupOne(*m_db, block_index, entry)) {
         return false;
     }
 
+    if (block_index->nHeight % CFCHECKPT_INTERVAL == 0) {
+        // Add to the headers cache
+        LOCK(m_cs_headers_cache);
+        if (m_headers_cache.size() < CF_HEADERS_CACHE_MAX_SZ) {
```

Comment:
doesn't this statically cap (and freeze forever) the cache after inserting 2000 entries? Wouldn't it make sense to have some sort of invalidation logic? (I can't come up with a good one though)

Reply:
If someone can mine enough blocks to fill up this cache, we have other problems. I think an adversary needs to mine blocks in the order of the 2000th triangle number to fill this up.

And when this cache is full due to natural reasons (in 20 years), disks will be fast enough that we won't need a cache.

Reply:
Maybe disks are fast enough or smart enough with caching today already, so it might not be needed after all. A benchmark would be nice...

Reply:
Oh. You'r right. 
I missed the `% CFCHECKPT_INTERVAL == 0`.

Reply:
Remind me to change this to 4000 in 2026 :)

I agree that a benchmark would be nice.

Reply:
```cpp
// TODO change to 4000 in 2026
```

;)

Reply:
oops sorry, I meant 2046